### PR TITLE
prov/verbs: Set QP min_rnr_timer to a good value to fix performance issues

### PR DESCRIPTION
**Pull request by a-ilango**
This is needed to avoid performance issues in streaming and bandwidth tests.
In those tests the sender could easily overrun the receiver and min_rnr_timer
determines how quick the receiver would communicate to the sender about lack
of posted receives for incoming sends requests.

Also added an env variable to allow user to modify default min_rnr_timer value.

### CODE REVIEW

File: prov/verbs/src/verbs_eq.c [hunk 1]
```diff
@@ -164,6 +165,9 @@ fi_ibv_eq_cm_process_event(struct fi_ibv_eq *eq, struct rdma_cm_event *cma_event
 	case RDMA_CM_EVENT_ESTABLISHED:
 		*event = FI_CONNECTED;
 		entry->info = NULL;
+		ret = fi_ibv_modify_qp(cma_event->id->qp);
```
Comment (gladkovdmitry17) [inline prov/verbs/src/verbs_eq.c]: If I correctly know, EQ isn't required in case of RDM EP type.
Maybe move it to fi_ibv_rdm_pack_cm_params for RDM type,  fi_ibv_msg_ep_connect and fi_ibv_msg_ep_accept for MSG type.
What do you think regarding proposed changes?

File: prov/verbs/src/fi_verbs.c [hunk 2]
```diff
@@ -390,6 +392,53 @@ static int fi_ibv_get_param_int(char *param_name, char *param_str,
 	return 0;
 }
 
+int fi_ibv_modify_qp(struct ibv_qp *qp)
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Rename to something like: fi_ibv_set_rnr_timer()

File: prov/verbs/src/fi_verbs.c [hunk 3]
```diff
@@ -390,6 +392,53 @@ static int fi_ibv_get_param_int(char *param_name, char *param_str,
...
+	int ret;
+
+	attr.min_rnr_timer = verbs_min_rnr_timer;
+
+	ret = ibv_modify_qp(qp, &attr, IBV_QP_MIN_RNR_TIMER);
+	if (ret) {
+		VERBS_WARN(FI_LOG_EQ, "Unable to modify QP attribute\n");
+		return ret;
+	}
+	ret = fi_ibv_dbg_query_qp_attr(qp);
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: This cause a build error.

File: prov/verbs/src/fi_verbs.c [hunk 4]
```diff
@@ -390,6 +392,53 @@ static int fi_ibv_get_param_int(char *param_name, char *param_str,
...
+		     qp_attr.min_rnr_timer, qp_attr.timeout, qp_attr.retry_cnt,
+		     qp_attr.rnr_retry);
+	return 0;
+}
+#else
+int fi_ibv_dbg_qp_attr(struct ibv_qp *qp)
+{
+	return 0;
+}
+#endif
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Name changed from debug to free build.  Please move up before use and make static.

File: prov/verbs/src/fi_verbs.c [hunk 5]
```diff
@@ -445,5 +494,15 @@ VERBS_INI
...
 
+	if (fi_ibv_get_param_int("min_rnr_timer", "Set min_rnr_timer QP "
+				 "attribute (0 - 31)", &verbs_min_rnr_timer))
+		return NULL;
+
+	if (verbs_min_rnr_timer > 31) {
+		VERBS_WARN(FI_LOG_CORE, "Invalid value for min_rnr_timer."
+			   "Valid range: 0 - 31 \n");
+		return NULL;
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Separate from this patch, I wonder if the code shouldn't return on error, but instead try to continue and read other parameters.

File: prov/verbs/src/verbs_eq.c (hunk 1 above)
Comment (shefty) [inline prov/verbs/src/verbs_eq.c]: The intent is to set this after the connection has been established.

File: prov/verbs/src/verbs_msg_ep.c [hunk 6]
```diff
@@ -30,6 +30,8 @@
  * SOFTWARE.
  */
 
+#include <inttypes.h>
```
Comment (shefty) [inline prov/verbs/src/verbs_msg_ep.c]: Looks like an unnecessary change

File: prov/verbs/src/fi_verbs.c (hunk 5 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: We can't proceed with an invalid value for this attribute. But if we want to check other parameters as well before returning an error, we can do that.

File: prov/verbs/src/fi_verbs.c (hunk 4 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Yes. I missed the rename for the release build.

File: prov/verbs/src/verbs_eq.c (hunk 1 above)
Comment (a-ilango) [inline prov/verbs/src/verbs_eq.c]: Yes. Since we can change the QP attributes after the connection is established we can do it in one place. I believe verbs/RDM may not need this change as it has unexpected queue to handle incoming sends that don't have posted recvs.

File: prov/verbs/src/fi_verbs.c [hunk 7]
```diff
@@ -390,6 +392,53 @@ static int fi_ibv_get_param_int(char *param_name, char *param_str,
...
+	struct ibv_qp_attr qp_attr = { 0 };
+	int ret;
+
+	ret = ibv_query_qp(qp, &qp_attr, IBV_QP_TIMEOUT | IBV_QP_RETRY_CNT |
+			   IBV_QP_RNR_RETRY | IBV_QP_MIN_RNR_TIMER, &attr);
+	if (ret) {
+		VERBS_WARN(FI_LOG_EP_CTRL, "Unable to query QP\n");
+		return ret;
+	}
+	FI_DBG_TRACE(&fi_ibv_prov, FI_LOG_EP_CTRL, "QP attributes: "
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: We don't have a VERBS_DBG_TRACE. I wanted this to be at TRACE level since it's likely to get printed only once and not on every send/recv.

Also, I'm agnostic between FI_DBG* and VERBS_DBG* so I'm gonna leave it as it is.

File: prov/verbs/src/verbs_eq.c (hunk 1 above)
Comment (shefty) [inline prov/verbs/src/verbs_eq.c]: Even the unexpected queue requires posting receives to the QP.   I think the change can apply there as well.  Not sure where *there* is though.

File: prov/verbs/src/fi_verbs.c (hunk 5 above)
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Yes, I meant continue checking other parameters.  Currently this function aborts checks if it gets any error, leaving other parameters set to their default.  It would be better to log an error and continue checking.

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Hmm... that seems a little harsh for an error reading an environment variable.  Logging an error message, but continuing to run seems like a friendlier option.

File: prov/verbs/src/verbs_eq.c (hunk 1 above)
Comment (a-ilango) [inline prov/verbs/src/verbs_eq.c]: > Even the unexpected queue requires posting receives to the QP.

Yes but this change fixes the issue where the sender could overrun receive queue in the QP. I think that might be rare when there is an unexpected queue. Also, I'm not sure if verbs/RDM protocol has any software flow control to handle that.

Nevertheless, this change can be added to verbs/RDM as well to avoid any potential queue overruns.

File: prov/verbs/src/fi_verbs.c (hunk 5 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Yes, but I think the choice is between unexpected behavior (using default value when user had specified a value, albeit erroneous) vs returning an error. I'm leaning towards the latter but I'll let you make the call. :)

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Most of these are changing some default value, that is later adjusted based on hardware limits anyway.  So my vote is to attempt to continue, rather than having an error be a showstopper.

**Comment by a-ilango** [comment]
Updated the PR based on suggestions.

File: prov/verbs/src/verbs_eq.c [hunk 8]
```diff
@@ -164,6 +165,9 @@ fi_ibv_eq_cm_process_event(struct fi_ibv_eq *eq, struct rdma_cm_event *cma_event
 	case RDMA_CM_EVENT_ESTABLISHED:
 		*event = FI_CONNECTED;
 		entry->info = NULL;
+		ret = fi_ibv_set_rnr_timer(cma_event->id->qp);
```
Comment (shefty) [inline prov/verbs/src/verbs_eq.c]: rnr timer is an IB specific setting.  I'm not sure iwarp devices won't fail this call.

File: prov/verbs/src/verbs_eq.c [hunk 9]
```diff
@@ -164,6 +165,12 @@ fi_ibv_eq_cm_process_event(struct fi_ibv_eq *eq, struct rdma_cm_event *cma_event
 	case RDMA_CM_EVENT_ESTABLISHED:
 		*event = FI_CONNECTED;
 		entry->info = NULL;
+		if (cma_event->id->qp->context->device->transport_type !=
+		    IBV_TRANSPORT_IWARP) {
```
Comment (a-ilango) [inline prov/verbs/src/verbs_eq.c]: Added a check for IWARP transport.

**Comment by shefty** [comment]
Merging, but I would look at letting the user specify a value that actually makes sense.  No one intuitively knows what the values 0-31 represent, what a default of 12 means, or even if a smaller number or higher number makes the timeout smaller or bigger.

**Comment by a-ilango** [comment]
The description of values for min_rnr_timer are present in libibverbs man pages. The user could look up those. We can add a pointer to them in fi_verbs man page.

**Comment by shefty** [comment]
I think we can be friendlier and let the user specify the retry value in microseconds or milliseconds, and convert accordingly.  Plus maybe provide more details on what an "rnr" is, so they might have a clue why changing it could matter.  :)

**Comment by a-ilango** [comment]
Sounds good. I'll do a follow up PR.  👍
